### ISSUE TITLE
Make all transforms able to use primitive type attributes

### ISSUE THREAD

**Issue by GFabien** [fepegar/torchio#159]
**🚀 Feature / Improvement**
Hi @fepegar! It would be great to have all transforms able to use primitive type attributes to handle most use cases but for now `image_interpolation` for a few transforms must be one of the types in `torchio.Interpolation` and `HistogramStandardization` requires numpy arrays and extra work from the user to apply it to his dataset compared to most transforms (it may be related to https://github.com/fepegar/torchio/issues/102).

**Motivation**

The reason why I'm pointing this out is that I would like to define my whole pipeline (i.e. loading data, generating dataset with transforms, loading model, training and evaluating) based on configuration files. The benefits will be to have the same code for all experiments and to have files describing precisely what was done during the experiments. However, the problem is that it makes it hard to use attributes that are not primitive type attributes.

**Implementation**

For the problem with `image_interpolation`, it seems really easy to solve it as all augmentation transform using it are calling `parse_interpolation` from the base `RandomTransform` class (except `RandomMotion` which is not doing any parsing, probably a mistake) and the only other transform using it is `Resample` that implements its own `parse_interpolation` method. What I suggest is to modify the `parse_interpolation` methods to be able to accept either a `torchio.Interpolation` or either a string or an integer.

For the `HistogramStandardization` I think it would be interesting to have an option (maybe in another class) to use it as a "regular transform" when creating the dataset by computing the histograms on all `INTENSITY` images but I guess the difficulty here is that the standard way of applying transform is subject based and transforms do not see the full dataset.

What do you think of all of this @fepegar? If you think it's interesting I can provide PRs but I would probably need help to deal with `HistogramStandardization`.

**Comment by fepegar** [comment]
Hi @GFabien,

I think that makes a lot of sense. Enums are nice, but I've never been too happy with the implementation. Maybe the interpolation can be retrieved from the `Interpolation` class with e.g. `getattr(Interpolation, 'nearest'.upper())`. Also, I wonder if we should deprecate the use of `Interpolation` as an argument.

Maybe the way to go for `HistogramStandardization` is to make it able to get either a dictionary or a path. If it's a dictionary, the keys are the image names and the values are NumPy arrays or paths to `.npy` files. If it's a path, it should be a `.pth` to which the dict must have been saved. What do you think?

**Comment by GFabien** [comment]
Yes I think `getattr(Interpolation, 'nearest'.upper())` would be a nice way to do it. I suggested to be able to use an integer as well because in the `Resample` class there's a clear mapping between the different interpolation names and the interpolation order so that can also be an option. What do you think about it? I think you're right, deprecating the use of `Interpolation` as an argument is probably a good idea.

I think that would solve my issue as having primitive attributes but I was suggesting something different that would impact more the way we would use `HistogramStandardization`. Indeed I am not sure I get why the class is "flexible". Indeed, I have the feeling that the "95% time usage" (but maybe I'm wrong about that) is that you have your dataset and you want to apply histogram standardization to your whole dataset by computing the histograms of all your images in your dataset and then apply the standardization to your dataset. For now, you first need to call `HistogramStandardization.train()` and then use its result to apply `HistogramStandardization`. What I suggest is that the default would be to call `HistogramStandardization` with no arguments as a "regular transform" and it would return the standardized samples. You could also give an optional `output_path` parameter if you want to save the landmarks for further use or reapplying the transform to the same dataset without having to compute them again.

What do you think about it? I know that it will introduce breaking changes for people who are currently using this transform so if we want to go this way we will have to think about how to handle this issue.

**Comment by romainVala** [comment]
Hi there
about HistogramStandardization
I thing that this flexibility is important from a computation time point of view: getting the histogram for the all dataset can be very long.

Having only one call, as you propose, may be a good idea, (for simplicity on small dataset) but there need to be the possibility to save and load from file. 
So the way I see, will be to have an output_file argument and then
if the file does not exist -> do the train and save it
if it exists -> just do the load

A may be slightly off topic point is that I would like to have the histogram normalization as a Random transform for data augmentation
In this case I guess we would need a new class RandomHistogramNormalization but it will share the apply function from the HistogramStandartization class

**Comment by fepegar** [comment]
@GFabien 

> Yes I think getattr(Interpolation, 'nearest'.upper()) would be a nice way to do it. I suggested to be able to use an integer as well because in the Resample class there's a clear mapping between the different interpolation names and the interpolation order so that can also be an option. What do you think about it?

I prefer to use strings for interpolation to be more explicit. `Resample` converts to integer because NiBabel expects an integer representing the spline order. If I understand correctly, a spline of order 0 is equivalent to nearest neighbor, 1 is linear, etc.
The supported values can be retrieved as `supported_values = [key.name.lower() for key in Interpolation]`.

> I have the feeling that the "95% time usage" (but maybe I'm wrong about that) is that you have your dataset and you want to apply histogram standardization to your whole dataset by computing the histograms of all your images in your dataset and then apply the standardization to your dataset.

I'm not sure about that. I think a more common approach is to precompute the landmarks from a number of images and then use the landmark for preprocessing during training, as it's a very fast operation. If you precompute your standardized images, you'll need to save in disk the whole dataset, possibly in `float`, which seems quite inefficient.


@romainVala 

> if the file does not exist -> do the train and save it
if it exists -> just do the load

This is what NiftyNet does, but I remember it caused me some problems. I'd rather leave the user the responsibility of checking whether the file exists or not.

> A may be slightly off topic point is that I would like to have the histogram normalization as a Random transform for data augmentation

Yes, this is a different issue (I'm not sure what you want to do, but note that *all* transforms take a `p` argument and that we now have `OneOf`).

**Comment by GFabien** [comment]
> I prefer to use strings for interpolation to be more explicit. Resample converts to integer because NiBabel expects an integer representing the spline order. If I understand correctly, a spline of order 0 is equivalent to nearest neighbor, 1 is linear, etc.
The supported values can be retrieved as supported_values = [key.name.lower() for key in Interpolation].

Ok perfect. I totally agree!

> I'm not sure about that. I think a more common approach is to precompute the landmarks from a number of images and then use the landmark for preprocessing during training, as it's a very fast operation. If you precompute your standardized images, you'll need to save in disk the whole dataset, possibly in float, which seems quite inefficient.

Ok, I guess it is done only using a subset of the dataset and probably not taking samples randomly as you want to have some samples from each different scanner for example. I now better understand why it is done this way and it makes sense. Thanks!

**Comment by fepegar** [comment]
I typically use the only training set to train the histogram, to avoid [data leakage](https://machinelearningmastery.com/data-leakage-machine-learning/). Then you use the landmarks to standardize training, validation and test sets during training.

**Comment by GFabien** [comment]
Yes sure, but are you using your whole training set to train the histogram then?

**Comment by fepegar** [comment]
Let's say I have 1000 subjects. I randomly split them into 800 for training, 100 for validation, 100 for testing. I'd use the 800 to train the histogram.

**Comment by romainVala** [comment]
I am not convince that  taking the all population may induce data leakage since it is used only to compute a mean histogram

but this is the way scaller are used in sklearn so I gess it make sense, 

but I am alway interested to better understand

**Comment by fepegar** [comment]
You're using information from the outside the training set to train the model, that's definitely data leakage :) There's more info in the link I posted above. That's why sklearn does it that way too.

**Comment by fepegar** [comment]
Reopening as the `HistogramStandardization` case hasn't been handled yet.

**Comment by fepegar** [comment]
Fixed by @GFabien in #160 #164.
